staging/src/k8s.io/client-go/rest/request.go

```diff
@@ -691,6 +694,17 @@ func (r *Request) Stream() (io.ReadCloser, error) {
 	}
 }
 
+func (r *Request) StreamWithPing(ctx context.Context, pingInterval time.Duration) (io.ReadCloser, error) {
+	if ctx != nil && pingInterval > 0 {
+		client := r.client
+		if client == nil {
+			client = http.DefaultClient
+		}
+		keepalive.KeepAliveWithPinger(ctx, keepalivehttp.NewHttpPinger(client.(*http.Client), r.URL().Scheme, r.URL().Host), pingInterval)
```

Comment:
> I wouldn't expect making another request using the same client to necessarily keep alive an existing streaming connection

I will investigate more on this. On my environment for reproducing the L4 issue I can see that the fix works, but I might be relying on specific behavior.

Reply:
it might work for a http/2 connection which multiplexes requests over a single connection, but that doesn't seem good to rely on

Reply:
Yes, on the spdy part of the patch I'm confident, because we are sending a native ping frame over the multiplexed connection.

The other two methods is what I am not that sure about, and that I might be extrapolating specific behavior.

Reply:
After some investigation I have seen that the TCP socket is always reused, if on the same process of course. Golang implements the HTTP/1.1 pipelining and this is what allows the TCP socket to be reused, even when one of the requests is still streaming (however, I don't fully understand how this can work -- keep the stream open while performing other requests).

I have used different methods to confirm this: on one side, the `haproxy` stats socket, that always did show a single connection per process, `netstat` on my laptop that again did show only one port per process and also a slight variation of the PR with golang's client tracing to check what connections were retrieved from the pool (and they were all the time being marked as reused, same connection object address all the time).

All this being said, I will investigate more and fully confirm if this is the case; but experts feedback for confirmation would be awesome.

Reply:
I see now, so the connection is upgraded to HTTP/2.